File: src/main/java/org/labun/jooq/generator/GenerationTool.java
package org.labun.jooq.generator;

import lombok.SneakyThrows;
import org.jooq.util.Database;
import org.jooq.util.jaxb.Catalog;
import org.jooq.util.jaxb.Schema;
import org.labun.jooq.generator.config.DatabaseConfig;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.Arrays;

/**
 * @author Konstantin Labun
 */
public class GenerationTool {

    @SneakyThrows
    public static void generate(Generator generator) {
        DatabaseConfig db = generator.configuration().database();
        Class.forName(db.driverClass());

        try (Connection connection = DriverManager.getConnection(
                db.jdbcUrl(),
                db.username(),
                db.password())) {
            assertConnectionSuccess(connection);
            generator.generate(createDbMeta(db, connection));
        }
    }

    private static Database createDbMeta(DatabaseConfig db, Connection connection) throws InstantiationException, IllegalAccessException, ClassNotFoundException {
        Database dbMeta = (Database) Class.forName(db.dbMetaClass()).newInstance();
        configureDefaults(dbMeta);
        dbMeta.setConnection(connection);
        return dbMeta;
    }

    private static void assertConnectionSuccess(Connection connection) throws SQLException {
        connection.createStatement().execute("SELECT 1");
    }

    private static void configureDefaults(Database db) {
        db.setConfiguredSchemata(Arrays.asList(new Schema().withInputSchema("")));
        db.setConfiguredCatalogs(Arrays.asList(new Catalog().withInputCatalog("")));
        db.setIncludes(new String[]{".*"});
        db.setConfiguredCustomTypes(Arrays.asList());
        db.setConfiguredEnumTypes(Arrays.asList());
        db.setConfiguredForcedTypes(Arrays.asList());
    }

    private GenerationTool() {

    }
}

File: src/main/kotlin/org/labun/jooq/generator/GenerationTool.kt
package org.labun.jooq.generator

import org.jooq.util.Database
import org.jooq.util.jaxb.Catalog
import org.jooq.util.jaxb.Schema
import org.labun.jooq.generator.config.DatabaseConfig
import java.sql.Connection
import java.sql.DriverManager
import java.util.*

/**
 * @author Konstantin Labun
 */
object GenerationTool {

  fun generate(generator: Generator) {
    val db = generator.configuration.database
    Class.forName(db.driverClass)

    DriverManager.getConnection(db.jdbcUrl, db.username, db.password)
        .use { connection ->
          assertConnectionSuccess(connection)
          generator.generate(createDbMeta(db, connection))
        }
  }

  private fun createDbMeta(db: DatabaseConfig, connection: Connection): Database {
    val dbMeta = Class.forName(db.dbMetaClass).newInstance() as Database
    configureDefaults(dbMeta)
    dbMeta.connection = connection
    return dbMeta
  }

  private fun assertConnectionSuccess(connection: Connection) {
    connection.createStatement().execute("SELECT 1")
  }

  private fun configureDefaults(db: Database) {
    db.setConfiguredSchemata(Arrays.asList(Schema().withInputSchema("")))
    db.setConfiguredCatalogs(Arrays.asList(Catalog().withInputCatalog("")))
    db.includes = arrayOf(".*")
    db.configuredCustomTypes = listOf()
    db.configuredEnumTypes = listOf()
    db.configuredForcedTypes = listOf()
  }
}
